akka-http-jackson unable to unmarshall JSON array?

I have a simple program (inspired by a sample from another blog) that receives data to a POST request at JSON, and works fine when I use the json4s variant. However when I run the same requests into a version of the app that uses the Hakka-http-json I see an unmarshalling exception when unmarshalling a JSON array.
I'd expect equivalent behaviour in both examples, so I suspect something is missing in either my code, documentation, or elsewhere?

Json4s:
$ curl -X POST -H 'Content-Type: application/json' -d '[{"name": "Julie-ann", "age": 21}, {"name": "George", "age": 22}]' http://localhost:8080/customers
"Added 2 customers\n
$ curl http://localhost:8080/customers
[{"name":"Julie-ann","age":21},{"name":"George","age":22}]

Jackson:
$ curl -X POST -H 'Content-Type: application/json' -d '[{"name": "Julie-ann", "age": 21}, {"name": "George", "age": 22}]' http://localhost:8080/customers
There was an internal server error

 Server:
[info] Running com.example.JacksonHttpApp 
Press RETURN to stop...
[INFO] [08/09/2017 15:54:40.459] [JacksonHttpApp-akka.actor.default-dispatcher-3] [akka.actor.ActorSystemImpl(JacksonHttpApp)] Server online at http://localhost:8080/
[ERROR] [08/09/2017 15:54:44.851] [JacksonHttpApp-akka.actor.default-dispatcher-6] [akka.actor.ActorSystemImpl(JacksonHttpApp)] Error during processing of request: 'scala.collection.immutable.Map$Map2 cannot be cast to com.example.JacksonHttpApp$Customer'. Completing with 500 Internal Server Error response. To change default exception handling behavior, provide a custom ExceptionHandler.
java.lang.ClassCastException: scala.collection.immutable.Map$Map2 cannot be cast to com.example.JacksonHttpApp$Customer
	at scala.collection.immutable.List.foreach(List.scala:389)
	at com.example.JacksonHttpApp$.$anonfun$routes$4(AkkaJacksonParsing.scala:27)
	at akka.http.scaladsl.server.directives.RouteDirectives.$anonfun$complete$1(RouteDirectives.scala:47)
...

Progs attached here...
[progs.zip](https://github.com/hseeberger/akka-http-json/files/1211862/progs.zip)

--- a/akka-http-jackson/src/main/scala/de/heikoseeberger/akkahttpjackson/JacksonSupport.scala
+++ b/akka-http-jackson/src/main/scala/de/heikoseeberger/akkahttpjackson/JacksonSupport.scala
@@ -16,16 +16,20 @@
 
 package de.heikoseeberger.akkahttpjackson
 
+import java.lang.reflect.{ ParameterizedType, Type => JType }
+
 import akka.http.javadsl.marshallers.jackson.Jackson
 import akka.http.scaladsl.marshalling._
 import akka.http.scaladsl.model.ContentTypeRange
 import akka.http.scaladsl.model.MediaTypes.`application/json`
 import akka.http.scaladsl.unmarshalling.{ FromEntityUnmarshaller, Unmarshaller }
 import akka.util.ByteString
+import com.fasterxml.jackson.core.`type`.TypeReference
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.fasterxml.jackson.module.scala.DefaultScalaModule
+
 import scala.collection.immutable.Seq
-import scala.reflect.ClassTag
+import scala.reflect.runtime.universe._
 
 /**
   * Automatic to and from JSON marshalling/unmarshalling usung an in-scope Jackon's ObjectMapper
@@ -53,15 +57,36 @@ trait JacksonSupport {
         case (data, charset)       => data.decodeString(charset.nioCharset.name)
       }
 
+  private def typeReference[T: TypeTag] = {
+    val t      = typeTag[T]
+    val mirror = t.mirror
+    def mapType(t: Type): JType =
+      if (t.typeArgs.isEmpty) {
+        mirror.runtimeClass(t)
+      } else {
+        new ParameterizedType {
+          def getRawType = mirror.runtimeClass(t)
+
+          def getActualTypeArguments = t.typeArgs.map(mapType).toArray
+
+          def getOwnerType = null
+        }
+      }
+
+    new TypeReference[T] {
+      override def getType = mapType(t.tpe)
+    }
+  }
+
   /**
     * HTTP entity => `A`
     */
   implicit def unmarshaller[A](
-      implicit ct: ClassTag[A],
+      implicit ct: TypeTag[A],
       objectMapper: ObjectMapper = defaultObjectMapper
   ): FromEntityUnmarshaller[A] =
     jsonStringUnmarshaller.map(
-      data => objectMapper.readValue(data, ct.runtimeClass).asInstanceOf[A]
+      data => objectMapper.readValue(data, typeReference[A]).asInstanceOf[A]
     )
 
   /**
@@ -71,4 +96,5 @@ trait JacksonSupport {
       implicit objectMapper: ObjectMapper = defaultObjectMapper
   ): ToEntityMarshaller[Object] =
     Jackson.marshaller[Object](objectMapper)
+
 }

--- a/akka-http-jackson/src/test/scala/de/heikoseeberger/akkahttpjackson/JacksonSupportSpec.scala
+++ b/akka-http-jackson/src/test/scala/de/heikoseeberger/akkahttpjackson/JacksonSupportSpec.scala
@@ -50,6 +50,14 @@ final class JacksonSupportSpec extends AsyncWordSpec with Matchers with BeforeAn
         .map(_ shouldBe foo)
     }
 
+    "should enable marshalling and unmarshalling of arrays of values" in {
+      val foo = Seq(Foo("bar"))
+      Marshal(foo)
+        .to[RequestEntity]
+        .flatMap(Unmarshal(_).to[Seq[Foo]])
+        .map(_ shouldBe foo)
+    }
+
     "provide proper error messages for requirement errors" in {
       val entity = HttpEntity(MediaTypes.`application/json`, """{ "bar": "baz" }""")
       Unmarshal(entity)

